url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/math/MathExpression.java

```diff
@@ -0,0 +1,615 @@
+ *     Martin - initial API and implementation
+ ******************************************************************************/
+package martin.math;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+
+import martin.quantum.tools.Tools;
+import martin.quantum.tools.Tuple;
+
+public class MathExpression implements MathsItem {
+	
+	public static int MAX_NUMBER_OF_SIMPLIFICAITON_STEPS = 50;//1000;
+	
+	final HashSet<HashSet<MathsItem>> items = new HashSet<HashSet<MathsItem>>();
+	public static boolean DEEP_SIMPLIFY = true;
+
+	@Override
+	public boolean hasNegativeSign() {
+		return false;
+	}
+
+	@Override
+	public void negate() {
+		for (HashSet<MathsItem> multiples : items)
+			for (MathsItem it : multiples) {
+				it.negate();
+				break;
+			}
+	}
+
+	@Override
+	public boolean isZero() {
+		return getValue(null).isZero();
+	}
+
+	@Override
+	public boolean isOne() {
+		return getValue(null).isOne();
+	}
+
+	@Override
+	public Complex getValue(final HashMap<String, Complex> rules) {
+		final Complex ans = new Complex(0, 0);
+		
+		for (HashSet<MathsItem> m : items) {
+			final Complex mult = new Complex(1, 0);
+			
+			for (MathsItem mi : m) {
+				final Complex smres = mi.getValue(rules);
+				mult.multiply(smres);
+			}
+			
+			ans.add(mult);
+		}
+		
+		return ans;
+	}
+	
+	/**
+	 * Simplifies 2*3*4 + 5*5 into 24 + 25
+	 */
+	private boolean simplifyMultiplicators() {
+		
+		boolean h = false;
+		
+		for (final HashSet<MathsItem> hs : items) {
+			
+			boolean happened = true;
+			
+			while (happened) {
+				
+				happened = false;
+			
+				final HashSet<MathsItem> toremove = new HashSet<MathsItem>();
+			
+				outer:for (final MathsItem m : hs)
+					for (final MathsItem m2 : hs)
+						if (m != m2 && m.multiply(m2)) { // should we have m.getClass().equals(m2.getClass())  here?
+							toremove.add(m2);
+							happened = true;
+							break outer;
+						}
+			
+				for (final MathsItem m : toremove)
+					hs.remove(m);
+				
+				h |= !toremove.isEmpty();
+			}
+			
+		}
+		
+		return h;
+	
+	}
+	
+	/**
+	 * Remove nested expressions
+	 */
+	private boolean removeNestedExps() {
+
+		boolean happend = false;
+		final HashSet<HashSet<MathsItem>> itclone = new HashSet<HashSet<MathsItem>>();
+		cloneItems(itclone);
+		items.clear();
+
+
+		outer : for (final HashSet<MathsItem> hs : itclone) {
+
+			if (hs.size() == 1) {
+
+				for (final MathsItem it : hs) {
+
+					if (it instanceof MathExpression) {
+
+						final MathExpression h = (MathExpression) it;
+						
+						for (final HashSet<MathsItem> hs2 : h.items) {
+							final HashSet<MathsItem> nmult = new HashSet<MathsItem>();
+							for (final MathsItem mit : hs2)
+								nmult.add(mit.clone());
+							items.add(nmult);
+						}
+						happend = true;
+						
+						continue outer;
+					} 
+
+					break;
+				}
+
+			}
+
+			items.add(hs);
+
+		}
+		
+		return happend;
+
+	}
+	
+	private boolean performAdditionOnSimpleExpressions() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.